### PR TITLE
Support for GROUPING SETS/CUBE/ROLLUP

### LINKED ISSUE #1327: implement grouping sets, cubes, and rollups
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

a challenge to design and implement `grouping set`s, `cube`s, and `rollup`s.

**Describe the solution you'd like**

similar to https://www.postgresql.org/docs/current/queries-table-expressions.html#QUERIES-GROUPING-SETS

**Describe alternatives you've considered**

**Additional context**

this would be a rather large piece of work so ideally can be split up into three subsets:

- [ ] update sqlparser https://github.com/sqlparser-rs/sqlparser-rs/pull/366
- [ ] grouping sets
- [ ] cubes
- [ ] rollups

### PR DESCRIPTION
**Pull request by thinkharderdev**
# Which issue does this PR close?

Closes #1327

TODO

- [x] Implement CUBE expansion
- [x] Implement ROLLUP expansion
- [x] Add SQL tests for CUBE/ROLLUP queries

Note that currently the sql parser doesn't seem to handle `GROUP BY GROUPING SETS ...` so we need to address that to test that explicitly. 

 # Rationale for this change

This PR adds support for GROUPING SETS (and special cases CUBE/ROLLUP) in the physical planner and execution plan.

# What changes are included in this PR?

There are three primary changes:

1. `AggregateExec` now takes a `Vec<Vec<(Arc<dyn PhysicalExpr>,String)>>` to represent grouping sets. A normal `GROUP BY` is just a special case. We expect the grouping sets to be "aligned". For example, for a SQL clause like `GROUP BY GROUPING SETS ((a),(b),(a,b))`, `AggregateExec` assumes that the planner will expand that to the grouping set `((a,NULL),(NULL,b),(a,b))`. We can't handle this in the execution plan because we don't have `ParialEq` for `PhysicalExpr`.
2. In `DefaultPhysicalPlanner` handle expanding and aligning grouping sets. This includes expanding CUBE/ROLLUP expressions and merging and aligning GROUPING SET expressions. 
3. Handle grouping sets correctly in optimizers. 

Also we include serialization for grouping set expression in `datafusion-proto`  

# Are there any user-facing changes?

SQL statements with CUBE/ROLLUP should now be supported. GROUPING SETS should also be supported but it seems like the sql parser is not handling them correctly. 

I don't think so.

### CODE REVIEW

**Comment by thinkharderdev** [comment]
cc @alamb @tustvold @Jimexist @yjshen @andygrove 

The part of this that I am least confident about is that I didn't break anything in any of the optimizers :). So if someone familiar with that code can review that part I would be very grateful.

**Comment by alamb** [comment]
Thanks @thinkharderdev  -- I'll try and find some time to review this over the weekend.

File: datafusion/core/src/physical_plan/aggregates/mod.rs [hunk 1]
```diff
@@ -90,18 +121,24 @@ impl AggregateExec {
...
         aggr_expr: Vec<Arc<dyn AggregateExpr>>,
         input: Arc<dyn ExecutionPlan>,
         input_schema: SchemaRef,
     ) -> Result<Self> {
-        let schema = create_schema(&input.schema(), &group_expr, &aggr_expr, mode)?;
+        let schema = create_schema(
+            &input.schema(),
+            &grouping_set.expr,
+            &aggr_expr,
+            grouping_set.groups.iter().flatten().any(|is_null| *is_null),
```
Comment (alamb) [inline datafusion/core/src/physical_plan/aggregates/mod.rs:133]: I wonder if extracting this code to a function such as  `GroupingSets::contains_null()` might make the code easier to read. The same comment applies to other places where `GroupingSets::groups` is referenced as well. 

Given the size of this PR already, definitely could be done as a follow on

File: datafusion/core/src/lib.rs [hunk 2]
```diff
@@ -204,6 +204,7 @@
 /// DataFusion crate version
 pub const DATAFUSION_VERSION: &str = env!("CARGO_PKG_VERSION");
 
+extern crate core;
```
Comment (alamb) [inline datafusion/core/src/lib.rs:207]: Why is this necessary?

File: datafusion/core/src/physical_plan/aggregates/mod.rs [hunk 3]
```diff
@@ -65,13 +66,43 @@ pub enum AggregateMode {
     FinalPartitioned,
 }
 
+/// Represents a GROUPING SET in the physical plan
```
Comment (alamb) [inline datafusion/core/src/physical_plan/aggregates/mod.rs:69]: 👍 
Thank you -- this is well commented and clear to understand

```suggestion
/// Represents `GROUP BY` clause in the plan plan (including the more general GROUPING SET)
```

I was thinking that many people encountering SQL will not be familiar with GROUPING SET but would be very familiar with GROUP BY -- what would you think about calling this structure `PhysicalGroupBy` or something to make the connection to `GROUPING` clearer?

File: datafusion/core/src/physical_plan/aggregates/row_hash.rs [hunk 4]
```diff
@@ -110,12 +111,15 @@ impl GroupedHashAggregateStreamV2 {
         // The expressions to evaluate the batch, one vec of expressions per aggregation.
         // Assume create_schema() always put group columns in front of aggr columns, we set
         // col_idx_base to group expression count.
-        let aggregate_expressions =
-            aggregates::aggregate_expressions(&aggr_expr, &mode, group_expr.len())?;
+        let aggregate_expressions = aggregates::aggregate_expressions(
```
Comment (alamb) [inline datafusion/core/src/physical_plan/aggregates/row_hash.rs:114]: FYI @yjshen  -- it would be really nice to try and consolidate `row_hash` and `hash` -- filed https://github.com/apache/arrow-datafusion/issues/2723 to track 👍

File: datafusion/expr/src/expr.rs [hunk 5]
```diff
@@ -270,6 +270,27 @@ pub enum GroupingSet {
...
+    /// is just the underlying list of exprs. For `GROUPING SET` we need to deduplicate
+    /// the exprs in the underlying sets.
+    pub fn distinct_expr(&self) -> Vec<Expr> {
+        match self {
+            GroupingSet::Rollup(exprs) => exprs.clone(),
+            GroupingSet::Cube(exprs) => exprs.clone(),
+            GroupingSet::GroupingSets(groups) => {
+                let mut exprs: Vec<Expr> = vec![];
+                for exp in groups.iter().flatten() {
+                    if !exprs.contains(exp) {
```
Comment (alamb) [inline datafusion/expr/src/expr.rs:284]: This is N^2 in the number of grouping sets -- probably not an issue, I just figured I would point it out

File: datafusion/expr/src/expr_fn.rs [hunk 6]
```diff
@@ -226,6 +227,21 @@ pub fn scalar_subquery(subquery: Arc<LogicalPlan>) -> Expr {
     Expr::ScalarSubquery(Subquery { subquery })
 }
 
+/// Create a grouping set
+pub fn grouping_set(exprs: Vec<Vec<Expr>>) -> Expr {
+    Expr::GroupingSet(GroupingSet::GroupingSets(exprs))
+}
+
+/// Create a grouping set
```
Comment (alamb) [inline datafusion/expr/src/expr_fn.rs:235]:
```suggestion
/// Create a grouping set for all combination of `exprs`
```

File: datafusion/expr/src/expr_fn.rs [hunk 7]
```diff
@@ -226,6 +227,21 @@ pub fn scalar_subquery(subquery: Arc<LogicalPlan>) -> Expr {
...
+pub fn cube(exprs: Vec<Expr>) -> Expr {
+    Expr::GroupingSet(GroupingSet::Cube(exprs))
+}
+
+/// Create a grouping set
```
Comment (alamb) [inline datafusion/expr/src/expr_fn.rs:240]:
```suggestion
/// Create a grouping set for rollup 
```

File: datafusion/core/tests/sql/aggregates.rs [hunk 8]
```diff
@@ -476,6 +476,205 @@ async fn csv_query_approx_percentile_cont() -> Result<()> {
...
+        "| e  | 4  | -53  | 13788                      |",
+        "| e  | 4  | 30   | -16110                     |",
+        "| e  | 4  | 73   | -22501                     |",
+        "| e  | 4  | 74   | -12612                     |",
+        "| e  | 4  | 96   | -30336                     |",
+        "| e  | 4  | 97   | -13181                     |",
+        "| e  | 4  |      | -16064.57142857143         |",
+        "| e  | 5  | -86  | 32514                      |",
+        "| e  | 5  | 64   | -26526                     |",
+        "| e  | 5  |      | 2994                       |",
```
Comment (alamb) [inline datafusion/core/tests/sql/aggregates.rs:669]: 👍

File: datafusion/expr/src/utils.rs [hunk 9]
```diff
@@ -45,6 +45,22 @@ pub fn exprlist_to_columns(expr: &[Expr], accum: &mut HashSet<Column>) -> Result
     Ok(())
 }
 
+/// Find all distinct exprs in a list of group by expressions. We assume that if the
+/// first element is a `GroupingSet` expression then that is the only expr.
```
Comment (alamb) [inline datafusion/expr/src/utils.rs:49]:
```suggestion
/// Find all distinct exprs in a list of group by expressions. If the
/// first element is a `GroupingSet` expression then it must be the only expr.
```

File: datafusion/optimizer/src/single_distinct_to_groupby.rs [hunk 10]
```diff
@@ -250,6 +280,29 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn single_distinct_and_grouping_set() -> Result<()> {
```
Comment (alamb) [inline datafusion/optimizer/src/single_distinct_to_groupby.rs:284]: Given there is special handling for CUBE and ROLLUP in this pass, I suggest test coverage for those cases too

**Review by alamb — COMMENTED** [review]
Thank you @thinkharderdev and @Tomczik76 -- this is super cool. I haven't made it all the way through yet but what I have reviewed is 👌 

I found the whitespace blind diff easier to review: https://github.com/apache/arrow-datafusion/pull/2716/files?w=1

cc @andygrove  @liukun4515

File: datafusion/core/src/lib.rs (hunk 2 above)
Comment (thinkharderdev) [reply datafusion/core/src/lib.rs]: It's not :) Not sure where it came from but removed now.

File: datafusion/expr/src/expr.rs (hunk 5 above)
Comment (thinkharderdev) [reply datafusion/expr/src/expr.rs:284]: Yeah, this is unfortunate

File: datafusion/optimizer/src/single_distinct_to_groupby.rs (hunk 10 above)
Comment (thinkharderdev) [reply datafusion/optimizer/src/single_distinct_to_groupby.rs:284]: Yeah, I think there is actually a bug in this. I'll work on a fix.

Comment (thinkharderdev) [reply datafusion/optimizer/src/single_distinct_to_groupby.rs:284]: Ok, this optimization is a bit more complicated for grouping sets. We need to create a separate alias for each group. For the moment I have just disabled the optimization for this case.

Comment (alamb) [reply datafusion/optimizer/src/single_distinct_to_groupby.rs:270]: I think disabling the optimization for grouping sets is a wise idea.

File: datafusion/core/src/physical_plan/planner.rs [hunk 11]
```diff
@@ -574,11 +569,12 @@ impl DefaultPhysicalPlanner {
 
                     // TODO: dictionary type not yet supported in Hash Repartition
                     let contains_dict = groups
+                        .expr
```
Comment (alamb) [inline datafusion/core/src/physical_plan/planner.rs:572]: I think it is a minor thing, but one might imagine keeping the fields of `PhysicalGroupBy` private and adding functions like `fn expr()` and `fn is_empty()` mostly as a way of additional documentation

File: datafusion/core/src/physical_plan/planner.rs [hunk 12]
```diff
@@ -1001,6 +1001,275 @@ impl DefaultPhysicalPlanner {
...
+
+    for expr_group in grouping_sets.iter() {
+        let mut group: Vec<bool> = Vec::with_capacity(expr_count);
+        for expr in all_exprs.iter() {
+            if expr_group.contains(expr) {
+                group.push(false);
+            } else {
+                group.push(true)
+            }
+        }
```
Comment (alamb) [inline datafusion/core/src/physical_plan/planner.rs:1120]: I don't think it matters, but you can probably express this in a functional style like:

```suggestion
        let group: Vec<bool> = all_exprs.iter()
          .map(expr_group.contains(expr))
          .collect();
```

File: datafusion/core/src/physical_plan/planner.rs [hunk 13]
```diff
@@ -1346,6 +1616,74 @@ mod tests {
...
 
+    #[tokio::test]
+    async fn test_create_cube_expr() -> Result<()> {
+        let logical_plan = test_csv_scan()
+            .await?
+            .project(vec![col("c1"), col("c2"), col("c3")])?
+            .aggregate(
+                vec![cube(vec![col("c1"), col("c2"), col("c3")])],
+                vec![sum(col("c2"))],
+            )?
```
Comment (alamb) [inline datafusion/core/src/physical_plan/planner.rs:1627]: I don't understand the need to creating the `aggregate` on the logical plan (as then new cube expressions are planned below). Can you simply use the output of the `project` plan?

The same question applies to the other plans below

File: datafusion/core/tests/sql/aggregates.rs [hunk 14]
```diff
@@ -1223,6 +1669,79 @@ async fn count_aggregated() -> Result<()> {
...
+        "| 3  | 6  | 1              |",
+        "| 3  | 7  | 1              |",
+        "| 3  | 8  | 1              |",
+        "| 3  | 9  | 1              |",
+        "+----+----+----------------+",
+    ];
+    assert_batches_sorted_eq!(expected, &results);
+    Ok(())
+}
+
```
Comment (alamb) [inline datafusion/core/tests/sql/aggregates.rs:1744]: I think the test coverage is quite good. Thank you

**Review by alamb — APPROVED** [review]
I think it looks great to me. Thanks again!

I had a few minor comments (e.g. some left over printlns) but all in all I think this one is good to go

File: datafusion/core/src/physical_plan/planner.rs (hunk 11 above)
Comment (thinkharderdev) [reply datafusion/core/src/physical_plan/planner.rs]: Yeah, I think that is a food idea. Fixed.

File: datafusion/core/src/physical_plan/aggregates/mod.rs [hunk 15]
```diff
@@ -117,14 +171,16 @@ impl AggregateExec {
 
     /// Grouping expressions
     pub fn group_expr(&self) -> &[(Arc<dyn PhysicalExpr>, String)] {
-        &self.group_expr
+        // TODO Is this right?
```
[the remark this answers is not in the document]
Comment (thinkharderdev) [reply datafusion/core/src/physical_plan/aggregates/mod.rs:207]: In this case it would still be correct right? The aggregate stats are only used if there is no group by which this would still represent correctly.

[the remark this answers is not in the document]
Comment (thinkharderdev) [reply datafusion/core/src/physical_plan/aggregates/mod.rs:207]: Or maybe this should just return `&PhysicalGroupBy` instead? I could see how this could lead to issues elsewhere if it is used for optimizations.

**Review by alamb — COMMENTED** [review]
👍

**Comment by alamb** [comment]
Looks great to me -- thanks again for all the work @thinkharderdev 🎉
